File: php/pool_test.php
<?php
/**
 * Simple test script for ShibuDb PHP connection pooling functionality
 *
 * Requires ShibuDb server running: sudo shibudb start 4444
 */

require_once __DIR__ . '/ShibuDbClient.php';

$passed = 0;
$failed = 0;

function testAssert($condition, $message, &$passed, &$failed)
{
    if ($condition) {
        echo "OK: $message\n";
        $passed++;
        return true;
    } else {
        echo "FAIL: $message\n";
        $failed++;
        return false;
    }
}

function testBasicPooling(&$passed, &$failed)
{
    echo "Testing basic connection pooling...\n";

    try {
        $pool = new ShibuDbConnectionPool(
            'localhost', 4444,
            'admin', 'admin',
            30, 1, 3
        );

        $client = $pool->getConnection();

        $response = $client->listSpaces();
        testAssert(
            isset($response['status']) || isset($response['spaces']),
            'List spaces via pool',
            $passed,
            $failed
        );

        $response = $client->createSpace('pool_test', 'key-value');
        testAssert(
            isset($response['status']) && ($response['status'] === 'OK' || $response['status'] === 'ERROR'),
            'Create space via pool',
            $passed,
            $failed
        );

        $response = $client->useSpace('pool_test');
        testAssert(
            isset($response['status']) && $response['status'] === 'OK',
            'Use space via pool',
            $passed,
            $failed
        );

        $response = $client->put('test_key', 'test_value');
        testAssert(
            isset($response['status']) && $response['status'] === 'OK',
            'Put via pool',
            $passed,
            $failed
        );

        $response = $client->get('test_key');
        testAssert(
            isset($response['status']) && $response['status'] === 'OK',
            'Get via pool',
            $passed,
            $failed
        );
        testAssert(
            isset($response['value']) && $response['value'] === 'test_value',
            'Get value matches',
            $passed,
            $failed
        );

        $pool->releaseConnection($client);

        $stats = $pool->getStats();
        testAssert(
            is_array($stats) && isset($stats['pool_size']),
            'Pool statistics',
            $passed,
            $failed
        );

        $pool->close();
        testAssert(true, 'Pool closed', $passed, $failed);

        return true;

    } catch (Exception $e) {
        echo "FAIL: Basic pooling test failed: " . $e->getMessage() . "\n";
        $failed++;
        return false;
    }
}

function testSequentialPoolUsage(&$passed, &$failed)
{
    echo "\nTesting sequential pool usage...\n";

    try {
        $pool = new ShibuDbConnectionPool(
            'localhost', 4444,
            'admin', 'admin',
            30, 2, 5
        );

        for ($i = 0; $i < 3; $i++) {
            $client = $pool->getConnection();
            $response = $client->listSpaces();
            testAssert(
                isset($response['status']) || isset($response['spaces']),
                "Sequential get $i",
                $passed,
                $failed
            );
            $pool->releaseConnection($client);
        }

        $pool->close();
        testAssert(true, 'Sequential pool test complete', $passed, $failed);

        return true;

    } catch (Exception $e) {
        echo "FAIL: Sequential pool test failed: " . $e->getMessage() . "\n";
        $failed++;
        return false;
    }
}

function testPoolErrorHandling(&$passed, &$failed)
{
    echo "\nTesting error handling...\n";

    // Test with invalid credentials - pool creation may succeed but getConnection may fail
    try {
        $pool = new ShibuDbConnectionPool(
            'localhost', 4444,
            'invalid_user',
            'invalid_password',
            30, 1, 2
        );

        try {
            $client = $pool->getConnection();
            $client->listSpaces();
            echo "FAIL: Expected authentication error\n";
            $failed++;
        } catch (ShibuDbAuthenticationError $e) {
            testAssert(true, 'Invalid credentials handled', $passed, $failed);
        } catch (ShibuDbPoolExhaustedError $e) {
            testAssert(true, 'Pool exhausted with invalid credentials', $passed, $failed);
        } catch (Exception $e) {
            testAssert(true, 'Error caught: ' . get_class($e), $passed, $failed);
        }

        $pool->close();

    } catch (Exception $e) {
        echo "OK: Pool with invalid credentials failed as expected\n";
        $passed++;
    }

    // Test with non-existent server
    try {
        $pool = new ShibuDbConnectionPool(
            'localhost', 9999,
            null, null,
            2, 1, 2
        );

        try {
            $client = $pool->getConnection();
            echo "FAIL: Expected connection error\n";
            $failed++;
        } catch (ShibuDbConnectionError $e) {
            testAssert(true, 'Non-existent server handled', $passed, $failed);
        } catch (ShibuDbPoolExhaustedError $e) {
            testAssert(true, 'Pool exhausted (connection failed)', $passed, $failed);
        } catch (Exception $e) {
            testAssert(true, 'Error caught', $passed, $failed);
        }

        $pool->close();

    } catch (ShibuDbConnectionError $e) {
        testAssert(true, 'Connection error on pool creation', $passed, $failed);
    } catch (Exception $e) {
        echo "OK: Non-existent server handled\n";
        $passed++;
    }
}

// Run tests
echo "ShibuDb PHP Connection Pooling Tests\n";
echo str_repeat('=', 50) . "\n";

testBasicPooling($passed, $failed);
testSequentialPoolUsage($passed, $failed);
testPoolErrorHandling($passed, $failed);

// Summary
echo "\n" . str_repeat('=', 50) . "\n";
echo "Results: $passed passed, $failed failed\n";

if ($failed === 0) {
    echo "All connection pooling tests passed.\n";
    exit(0);
} else {
    echo "Some tests failed.\n";
    exit(1);
}

File: python/pool_test.py
#!/usr/bin/env python3
"""
Simple test script for ShibuDb connection pooling functionality
"""

import time
from shibudb_client import create_connection_pool, ShibuDbError


def test_basic_pooling():
    """Test basic connection pooling functionality"""
    print("Testing basic connection pooling...")
    
    try:
        # Create a small connection pool
        pool = create_connection_pool(
            host="localhost",
            port=4444,
            username="admin",
            password="admin",
            min_size=1,
            max_size=3
        )
        
        # Test basic operations
        with pool.get_connection() as client:
            # List spaces
            response = client.list_spaces()
            print(f"✓ List spaces successful: {response.get('status', 'UNKNOWN')}")
            
            # Create a test space
            response = client.create_space("pool_test", "key-value")
            print(f"✓ Create space successful: {response.get('status', 'UNKNOWN')}")
            
            # Use the space
            response = client.use_space("pool_test")
            print(f"✓ Use space successful: {response.get('status', 'UNKNOWN')}")
            
            # Put and get data
            response = client.put("test_key", "test_value")
            print(f"✓ Put operation successful: {response.get('status', 'UNKNOWN')}")
            
            response = client.get("test_key")
            print(f"✓ Get operation successful: {response.get('status', 'UNKNOWN')}")
            print(f"  Retrieved value: {response.get('value', 'N/A')}")
        
        # Check pool statistics
        stats = pool.get_stats()
        print(f"✓ Pool statistics: {stats}")
        
        # Clean up
        pool.close()
        print("✓ Pool closed successfully")
        
        return True
        
    except Exception as e:
        print(f"✗ Basic pooling test failed: {e}")
        return False


def test_concurrent_pooling():
    """Test concurrent operations with connection pool"""
    print("\nTesting concurrent operations...")
    
    try:
        # Create a connection pool for concurrent operations
        pool = create_connection_pool(
            host="localhost",
            port=4444,
            username="admin",
            password="admin",
            min_size=2,
            max_size=5
        )
        
        def worker(worker_id):
            """Worker function for concurrent operations"""
            try:
                with pool.get_connection() as client:
                    # Create a unique space for this worker
                    space_name = f"concurrent_test_{worker_id}"
                    client.create_space(space_name, "key-value")
                    client.use_space(space_name)
                    
                    # Perform operations
                    for i in range(3):
                        key = f"worker_{worker_id}_key_{i}"
                        value = f"worker_{worker_id}_value_{i}"
                        client.put(key, value)
                    
                    # Verify data
                    response = client.get(f"worker_{worker_id}_key_0")
                    return f"Worker {worker_id}: {response.get('value', 'N/A')}"
                    
            except Exception as e:
                return f"Worker {worker_id} failed: {e}"
        
        # Run concurrent workers
        from concurrent.futures import ThreadPoolExecutor, as_completed
        
        with ThreadPoolExecutor(max_workers=3) as executor:
            futures = [executor.submit(worker, i) for i in range(3)]
            
            for future in as_completed(futures):
                result = future.result()
                print(f"✓ {result}")
        
        # Check final pool statistics
        stats = pool.get_stats()
        print(f"✓ Final pool statistics: {stats}")
        
        # Clean up
        pool.close()
        print("✓ Concurrent test completed successfully")
        
        return True
        
    except Exception as e:
        print(f"✗ Concurrent test failed: {e}")
        return False


def test_pool_error_handling():
    """Test error handling with connection pool"""
    print("\nTesting error handling...")
    
    try:
        # Test with invalid credentials
        pool = create_connection_pool(
            host="localhost",
            port=4444,
            username="invalid_user",
            password="invalid_password",
            min_size=1,
            max_size=2
        )
        
        try:
            with pool.get_connection() as client:
                client.list_spaces()
        except Exception as e:
            print(f"✓ Caught expected error with invalid credentials: {type(e).__name__}")
        
        pool.close()
        
        # Test with non-existent server
        pool = create_connection_pool(
            host="localhost",
            port=9999,
            min_size=1,
            max_size=2
        )
        
        try:
            with pool.get_connection() as client:
                client.list_spaces()
        except Exception as e:
            print(f"✓ Caught expected error with non-existent server: {type(e).__name__}")
        
        pool.close()
        print("✓ Error handling test completed successfully")
        
        return True
        
    except Exception as e:
        print(f"✗ Error handling test failed: {e}")
        return False


def main():
    """Run all connection pooling tests"""
    print("🚀 ShibuDb Connection Pooling Tests")
    print("=" * 50)
    
    tests = [
        ("Basic Pooling", test_basic_pooling),
        ("Concurrent Operations", test_concurrent_pooling),
        ("Error Handling", test_pool_error_handling)
    ]
    
    passed = 0
    total = len(tests)
    
    for test_name, test_func in tests:
        print(f"\n--- {test_name} ---")
        if test_func():
            passed += 1
            print(f"✓ {test_name} PASSED")
        else:
            print(f"✗ {test_name} FAILED")
    
    print(f"\n{'='*50}")
    print(f"Test Results: {passed}/{total} tests passed")
    
    if passed == total:
        print("🎉 All connection pooling tests passed!")
    else:
        print("❌ Some tests failed. Please check the output above.")
    
    return passed == total


if __name__ == "__main__":
    main() 